packages/contracts-bedrock/test/setup/Upgrade.s.sol

```diff
@@ -0,0 +1,113 @@
+import { IDisputeGameFactory } from "interfaces/dispute/IDisputeGameFactory.sol";
+import { IAddressManager } from "interfaces/legacy/IAddressManager.sol";
+
+/// @title Upgrade
+/// @notice This script is called by Setup.sol as a preparation step for the foundry test suite, and is run as an
+///         alternative to Deploy.s.sol, when `UPGRADE_TEST=true` is set in the env.
+///         Like Deploy.s.sol this script saves the system addresses to disk so that they can be read into memory later
+///         on, however rather than deploying new contracts from the local source code, it simply reads the addresses
+///         from the superchain-registry.
+///         Therefore this script can only be run against a fork of a production network which is listed in the
+///         superchain-registry.
+///         This contract must not have constructor logic because it is set into state using `etch`.
+contract Upgrade is Deployer {
+    using stdJson for string;
+
+    /// @notice Returns the base chain name to use for forking
+    /// @return The base chain name as a string
+    function upgradeBaseChain() internal view returns (string memory) {
+        return vm.envOr("UPGRADE_BASE_CHAIN", string("mainnet"));
+    }
+
+    /// @notice Returns the OP chain name to use for forking
+    /// @return The OP chain name as a string
+    function upgradeOpChain() internal view returns (string memory) {
+        return vm.envOr("UPGRADE_OP_CHAIN", string("op"));
+    }
+
+    /// @notice Reads a standard chains addresses from the superchain-registry and saves them to disk.
+    function run() public {
+        string memory superchainBasePath = "./lib/superchain-registry/superchain/configs/";
+
+        // Read the superchain config files
+        // During development of an upgrade which adds a new contract, the contract will not yet be present in the
+        // superchain-registry. In this case, the contract will be deployed by the upgrade process, and will need to
+        // be saved by after the call to opcm.upgrade().
+        // After the upgrade is complete, the superchain-registry will be updated and the contract will be present.
+        // At this point, the test will need to be updated to read the new contract from the superchain-registry using
+        // either the `saveProxyAndImpl` or `save` functions.
+        string memory superchainToml =
+            vm.readFile(string.concat(superchainBasePath, upgradeBaseChain(), "/superchain.toml"));
+        string memory opToml =
+            vm.readFile(string.concat(superchainBasePath, upgradeBaseChain(), "/", upgradeOpChain(), ".toml"));
+
+        // Superchain shared contracts
+        saveProxyAndImpl("SuperchainConfig", superchainToml, ".superchain_config_addr");
+        saveProxyAndImpl("ProtocolVersions", superchainToml, ".protocol_versions_addr");
+        save("OPContractsManager", vm.parseTomlAddress(superchainToml, ".op_contracts_manager_proxy_addr"));
```

Comment:
I would like to double check that the SCR is storing the correct OPCM addresses. I know there's been some churn and the canonical reference at this point is likely in op-deployer: https://github.com/ethereum-optimism/optimism/blob/develop/op-deployer/pkg/deployer/standard/standard.go#L202-L215

The OPCM currently stored in SCR is still proxied: https://etherscan.io/address/0x18CeC91779995AD14c880e4095456B9147160790#code 

More broadly, I think we need to all agree on a format that we want to store OPCM addresses in SCR. I've created an issue for this in SCR. I'm not sure where best to make sure it's being tracked. It's an important part of this work though: https://github.com/ethereum-optimism/superchain-registry/issues/806

Reply:
Good call, I've also noticed that.

I added it to my [master issue](https://github.com/ethereum-optimism/optimism/issues/13069) under Milestone:Mainnet.